### PR TITLE
Smartly choose a new layer name to prevent duplicate names

### LINKED ISSUE #3452: Automatic layer names should be unique
Tiled automatically names newly created layers, e.g. "Tile Layer 3". However, Tiled seems to use the number of layers of that type +1 as the number in this automatic name, which means layer names can repeat if, e.g. a user deleted some layers, or manually named some in that pattern. If I have two Tile Layers called "Tile Layer 1" and "Tile Layer 3", Tiled will suggest "Tile Layer 3" for the next one, which doesn't seem right and can lead to confusion.

Tiled should choose numbers for these layers such that the layer name is unique. A few options for this are possible:
1. Always the layer ID, instead of the layer count of that type of layer. This would be free since the layer ID is incremented anyway. However, this will mean that deleting layers or using multiple layer types will cause the names to be larger than expected in most cases, which will annoy some users (including me).
2. Do what Photoshop does: Find the largest layer name of that type, and add 1. If no such layer names exist, suggest "Tile Layer 1". So, if there are two layers, "Tile Layer 1" and "Tile Layer 5", Tiled would suggest "Tile Layer 6". This means if the user always gives custom names to all their layers after creating them, Tiled will suggest "Tile Layer 1" for every new layer.
3. Do a mix of Photoshop-type and current naming: Take the larger value of the layer count for that layer type, and the current layer names. So, if you have two tile layers, "Tile Layer 1" and "Other Tile Layer", Tiled will suggest "Tile Layer 3", but if it's "Tile Layer 5" and "Other Tile Layer", Tiled will suggest "Tile Layer 6".

I'd probably go with the second option, because I'm used to that style of layer naming from the art programs I use - Photoshop, Aseprite, and Krita all use that approach. However, I should mention that all of these programs name all layers with the same number sequence, regardless of type. In Krita, if you have "Paint Layer 1" and then add a new Filter Layer, it'll be called "Filter Layer 2", while in Aseprite and Photoshop, they're all just "Layer 1", "Layer 2", and so on. I think Tiled's independent numbering for each layer type makes sense for Tiled, and should be kept.

### PR DESCRIPTION
**Pull request by lhiginbotham**
Fixes #3452 by attempting to use the default "{Type} Layer {Number}" name, incrementing {Number} if any collisions occur.

### CODE REVIEW

**Review by bjorn — CHANGES_REQUESTED** [review]
Patch looks good, but I added some comments that need to be addressed. :-)

File: src/tiled/mapdocument.cpp [hunk 1]
```diff
@@ -1451,6 +1450,45 @@ void MapDocument::onLayerRemoved(Layer *layer)
...
+            break;
+        case Layer::TypeFlag::GroupLayerType:
+            parametricName = "Group Layer %1";
+            break;
+        default:
+            // invalid layer type
+            return {};
+    }
+
+    fillLayerNames(mMap->allLayers(layerType));
```
[the remark this answers is not in the document]
Comment (lhiginbotham) [reply src/tiled/mapdocument.cpp]: Done, thanks!

File: src/tiled/mapdocument.cpp [hunk 2]
```diff
@@ -1451,6 +1450,45 @@ void MapDocument::onLayerRemoved(Layer *layer)
     emit layerRemoved(layer);
 }
 
+QString MapDocument::getNewLayerName(Layer::TypeFlag layerType) const
```
[the remark this answers is not in the document]
Comment (lhiginbotham) [reply src/tiled/mapdocument.cpp]: Got it, thanks!

File: src/tiled/mapdocument.cpp [hunk 3]
```diff
@@ -1451,6 +1450,45 @@ void MapDocument::onLayerRemoved(Layer *layer)
...
+{
+    QSet<QString> layerNames;
+    const auto fillLayerNames = [&layerNames](auto layers) {
+        for (const auto& layer : layers)
+            layerNames.insert(layer->name());
+    };
+
+    const char* parametricName = nullptr;
```
[the remark this answers is not in the document]
Comment (lhiginbotham) [reply src/tiled/mapdocument.cpp]: Fixed 👍

File: src/tiled/mapdocument.cpp [hunk 4]
```diff
@@ -1451,6 +1450,45 @@ void MapDocument::onLayerRemoved(Layer *layer)
...
+            // invalid layer type
+            return {};
+    }
+
+    fillLayerNames(mMap->allLayers(layerType));
+
+    int number = 1;
+    auto candidateName = tr(parametricName).arg(number);
+    while (layerNames.contains(candidateName)) {
+        candidateName = tr(parametricName).arg(++number);
```
[the remark this answers is not in the document]
Comment (lhiginbotham) [reply src/tiled/mapdocument.cpp:1482]: Thanks for pointing this out to me, I don't work with Qt all that much

**Comment by lhiginbotham** [comment]
> Patch looks good, but I added some comments that need to be addressed. :-)

@bjorn - changes have been made, please check it out 😃

**Comment by bjorn** [comment]
@lhiginbotham Thanks! :-)

**Pull request merged**